File: src/android/MyTargetAdsPlugin.java
package io.luzh.cordova.plugin;

import android.util.Log;
import android.text.TextUtils;
import android.os.AsyncTask;
import android.view.View;
import android.view.ViewGroup;
import android.view.Gravity;
import android.view.ViewGroup.LayoutParams;
import android.widget.RelativeLayout;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

//import com.yandex.mobile.ads.banner.BannerAdView;
//import com.yandex.mobile.ads.banner.AdSize;
//import com.yandex.mobile.ads.banner.BannerAdEventListener;
//import com.yandex.mobile.ads.common.ImpressionData;
//import com.yandex.mobile.ads.common.InitializationListener;

import com.my.target.common.MyTargetPrivacy;
import com.my.target.ads.RewardedAd;
import com.my.target.ads.Reward;
import com.my.target.ads.InterstitialAd;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;

public class MyTargetAdsPlugin extends CordovaPlugin {

    private static final String TAG = "MY_TARGET_ADS";

    private static final String EVENT_INTERSTITIAL_LOADED = "interstitialLoaded";
    private static final String EVENT_INTERSTITIAL_FAILED_TO_LOAD = "interstitialFailedToLoad";
    private static final String EVENT_INTERSTITIAL_SHOWN = "interstitialShown";
    private static final String EVENT_INTERSTITIAL_CLOSED = "interstitialClosed";

    private static final String EVENT_REWARDED_VIDEO_LOADED = "rewardedVideoLoaded";
    private static final String EVENT_REWARDED_VIDEO_REWARDED = "rewardedVideoRewardReceived";
    private static final String EVENT_REWARDED_VIDEO_FAILED = "rewardedVideoFailed";
    private static final String EVENT_REWARDED_VIDEO_STARTED = "rewardedVideoStarted";
    private static final String EVENT_REWARDED_VIDEO_CLOSED = "rewardedVideoClosed";

    private static final String EVENT_BANNER_DID_LOAD = "bannerDidLoad";
    private static final String EVENT_BANNER_FAILED_TO_LOAD = "bannerFailedToLoad";
    private static final String EVENT_BANNER_DID_CLICK = "bannerDidClick";

    private RelativeLayout bannerContainerLayout;
    private CordovaWebView cordovaWebView;
    private ViewGroup parentLayout;
    private Boolean bannerLoaded = false;
    private Boolean bannerOverlap = false;
    private Boolean bannerAtTop = false;
    private String bannerSize = "BANNER_320x50";

    private RewardedAd mRewardedAd;
    private InterstitialAd mInterstitialAd;
    //private BannerAdView mBannerAdView;
    private Integer rewardedBlockId;
    private Integer interstitialBlockId;
    private Integer bannerBlockId;

//    private static Map<String, AdSize> bannerSizes;
//    static
//    {
//        bannerSizes = new HashMap<>();
//        bannerSizes.put("BANNER_240x400", AdSize.BANNER_240x400);
//        bannerSizes.put("BANNER_300x250", AdSize.BANNER_300x250);
//        bannerSizes.put("BANNER_300x300", AdSize.BANNER_300x300);
//        bannerSizes.put("BANNER_320x50", AdSize.BANNER_320x50);
//        bannerSizes.put("BANNER_320x100", AdSize.BANNER_320x100);
//        bannerSizes.put("BANNER_400x240", AdSize.BANNER_400x240);
//        bannerSizes.put("BANNER_728x90", AdSize.BANNER_728x90);
//    }

    @Override
    public boolean execute(String action, JSONArray args, final CallbackContext callbackContext) throws JSONException {

        if (action.equals("init")) {
            this.initAction(args, callbackContext);
            return true;
        }

        else if (action.equals("loadRewardedVideo")) {
            this.loadRewardedAction(args, callbackContext);
            return true;
        }

        else if (action.equals("showRewardedVideo")) {
            this.showRewardedVideoAction(args, callbackContext);
            return true;
        }

        else if (action.equals("loadBanner")) {
            this.loadBannerAction(args, callbackContext);
            return true;
        }

        else if (action.equals("showBanner")) {
            this.showBannerAction(args, callbackContext);
            return true;
        }

        else if (action.equals("hideBanner")) {
            this.hideBannerAction(args, callbackContext);
            return true;
        }

        else if (action.equals("loadInterstitial")) {
            this.loadInterstitialAction(args, callbackContext);
            return true;
        }

        else if (action.equals("showInterstitial")) {
            this.showInterstitialAction(args, callbackContext);
            return true;
        }

        else if (action.equals("setUserConsent")) {
            this.setUserConsentAction(args.getBoolean(0), callbackContext);
            return true;
        }

        return false;
    }

    /** --------------------------------------------------------------- */

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        cordovaWebView = webView;
        super.initialize(cordova, webView);
    }

    /** ----------------------- UTILS --------------------------- */

    private void emitWindowEvent(final String event) {
        final CordovaWebView view = this.webView;
        this.cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                view.loadUrl(String.format("javascript:cordova.fireWindowEvent('%s');", event));
            }
        });
    }

    private void emitWindowEvent(final String event, final JSONObject data) {
        final CordovaWebView view = this.webView;
        this.cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                view.loadUrl(String.format("javascript:cordova.fireWindowEvent('%s', %s);", event, data.toString()));
            }
        });
    }

    /** ----------------------- INITIALIZATION --------------------------- */

    /**
     * Intilization action Initializes Yandex Ads
     */
    private void initAction(JSONArray args, final CallbackContext callbackContext) throws JSONException {

        final MyTargetAdsPlugin self = this;
        rewardedBlockId = args.getInt(0);
        interstitialBlockId = args.getInt(1);
        bannerBlockId = args.getInt(2);

        JSONObject options = args.optJSONObject(3);
        if(options.has("overlap")) bannerOverlap = options.optBoolean("overlap");
        if(options.has("bannerAtTop")) bannerAtTop = options.optBoolean("bannerAtTop");
        if(options.has("bannerSize")) bannerSize = options.optString("bannerSize");
    }

    /**
     * Initializes Rewarded ads
     *
     * @todo Provide
     */
    private void initRewarded(Integer blockId) {

        final MyTargetAdsPlugin self = this;
        mRewardedAd = new RewardedAd(blockId, this.cordova.getActivity());
        mRewardedAd.setListener(new RewardedAd.RewardedAdListener()
        {
            @Override
            public void onLoad(RewardedAd ad)
            {
                Log.d(TAG, EVENT_REWARDED_VIDEO_LOADED);
                self.emitWindowEvent(EVENT_REWARDED_VIDEO_LOADED);
            }

            @Override
            public void onNoAd(String reason, RewardedAd ad)
            {
                Log.d(TAG, EVENT_REWARDED_VIDEO_FAILED + ": " + reason);
                self.emitWindowEvent(EVENT_REWARDED_VIDEO_FAILED);
            }

            @Override
            public void onClick(RewardedAd ad)
            {
            }

            @Override
            public void onDisplay(RewardedAd ad)
            {
                Log.d(TAG, EVENT_REWARDED_VIDEO_STARTED);
                self.emitWindowEvent(EVENT_REWARDED_VIDEO_STARTED);
            }

            @Override
            public void onDismiss(RewardedAd ad)
            {
                Log.d(TAG, EVENT_REWARDED_VIDEO_CLOSED);
                self.emitWindowEvent(EVENT_REWARDED_VIDEO_CLOSED);
            }

            @Override
            public void onReward(@NonNull Reward var1, @NonNull RewardedAd var2)
            {
                Log.d(TAG, EVENT_REWARDED_VIDEO_REWARDED);
                self.emitWindowEvent(EVENT_REWARDED_VIDEO_REWARDED);
            }
        });
    }

    private void initInterstitial(Integer blockId) {

        final MyTargetAdsPlugin self = this;

        mInterstitialAd = new InterstitialAd(blockId, this.cordova.getActivity());
        mInterstitialAd.setListener(new InterstitialAd.InterstitialAdListener()
        {
            @Override
            public void onLoad(InterstitialAd ad)
            {
                Log.d(TAG, EVENT_INTERSTITIAL_LOADED);
                self.emitWindowEvent(EVENT_INTERSTITIAL_LOADED);
            }

            @Override
            public void onNoAd(String reason, InterstitialAd ad)
            {
                Log.d(TAG, EVENT_INTERSTITIAL_FAILED_TO_LOAD + ": " + reason);
                self.emitWindowEvent(EVENT_INTERSTITIAL_FAILED_TO_LOAD);
            }

            @Override
            public void onClick(InterstitialAd ad)
            {
            }

            @Override
            public void onDisplay(InterstitialAd ad)
            {
                Log.d(TAG, EVENT_INTERSTITIAL_SHOWN);
                self.emitWindowEvent(EVENT_INTERSTITIAL_SHOWN);
            }

            @Override
            public void onDismiss(InterstitialAd ad)
            {
                Log.d(TAG, EVENT_INTERSTITIAL_CLOSED);
                self.emitWindowEvent(EVENT_INTERSTITIAL_CLOSED);
            }

            @Override
            public void onVideoCompleted(InterstitialAd ad)
            {
            }
        });
    }

    /**
     * ----------------------- VALIDATION INTEGRATION ---------------------------
     */

    /**
     * Validates integration action
     */
    private void validateIntegrationAction(JSONArray args, final CallbackContext callbackContext) {
    }

    /** ----------------------- REWARDED VIDEO --------------------------- */

    private void loadRewardedAction(JSONArray args, final CallbackContext callbackContext) {
        final MyTargetAdsPlugin self = this;
        cordova.getActivity().runOnUiThread(new Runnable() {
            public void run() {
                self.initRewarded(rewardedBlockId);
                mRewardedAd.load();
                callbackContext.success();
            }
        });
    }

    private void showRewardedVideoAction(JSONArray args, final CallbackContext callbackContext) throws JSONException {
        cordova.getActivity().runOnUiThread(new Runnable() {
            public void run() {
                mRewardedAd.show();
                callbackContext.success();
            }
        });
    }

    /** ----------------------- INTERSTITIAL --------------------------- */

    private void loadInterstitialAction(JSONArray args, final CallbackContext callbackContext) {
        final MyTargetAdsPlugin self = this;
        cordova.getActivity().runOnUiThread(new Runnable() {
            public void run() {
                self.initInterstitial(interstitialBlockId);
                mInterstitialAd.load();
                callbackContext.success();
            }
        });
    }

    private void showInterstitialAction(JSONArray args, final CallbackContext callbackContext) throws JSONException {
        cordova.getActivity().runOnUiThread(new Runnable() {
            public void run() {
                mInterstitialAd.show();
                callbackContext.success();
            }
        });
    }

    /** ----------------------- BANNER --------------------------- */
    private void showBannerAction(JSONArray args, final CallbackContext callbackContext) {
//        final MyTargetAdsPlugin self = this;
//
//        cordova.getActivity().runOnUiThread(new Runnable() {
//            public void run() {
//                if (mBannerAdView != null && bannerLoaded) {
//                    if (bannerOverlap) {
//                        RelativeLayout.LayoutParams params2 = new RelativeLayout.LayoutParams(
//                                RelativeLayout.LayoutParams.MATCH_PARENT,
//                                RelativeLayout.LayoutParams.WRAP_CONTENT);
//                        params2.addRule(bannerAtTop ? RelativeLayout.ALIGN_PARENT_TOP : RelativeLayout.ALIGN_PARENT_BOTTOM);
//
//                        RelativeLayout adViewLayout = new RelativeLayout(cordova.getActivity());
//                        RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.MATCH_PARENT);
//                        try {
//                            ((ViewGroup)(((View)webView.getClass().getMethod("getView").invoke(webView)).getParent())).addView(adViewLayout, params);
//                        } catch (Exception e) {
//                            ((ViewGroup) webView).addView(adViewLayout, params);
//                        }
//
//                        adViewLayout.addView(mBannerAdView, params2);
//                        adViewLayout.bringToFront();
//                    } else {
//                        parentLayout = (ViewGroup) cordovaWebView.getView().getParent();
//
//                        View view = cordovaWebView.getView();
//
//                        ViewGroup wvParentView = (ViewGroup) view.getParent();
//
//                        LinearLayout parentView = new LinearLayout(cordovaWebView.getContext());
//
//                        if (wvParentView != null && wvParentView != parentView) {
//                            ViewGroup rootView = (ViewGroup) (view.getParent());
//                            wvParentView.removeView(view);
//                            ((LinearLayout) parentView).setOrientation(LinearLayout.VERTICAL);
//                            parentView.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
//                                    ViewGroup.LayoutParams.MATCH_PARENT, 0.0F));
//                            view.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
//                                    ViewGroup.LayoutParams.MATCH_PARENT, 1.0F));
//                            parentView.addView(view);
//                            rootView.addView(parentView);
//                        }
//
//                        bannerContainerLayout = new RelativeLayout(self.cordova.getActivity());
//
//                        bannerContainerLayout.setGravity(Gravity.BOTTOM);
//
//                        RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(
//                                RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
//
//                        layoutParams.addRule(RelativeLayout.CENTER_HORIZONTAL);
//
//                        bannerContainerLayout.addView(mBannerAdView, layoutParams);
//
//                        mBannerAdView.setLayoutParams(layoutParams);
//
//                        if (bannerAtTop) {
//                            parentView.addView(bannerContainerLayout, 0);
//                        } else {
//                            parentView.addView(bannerContainerLayout);
//                        }
//                    }
//                }
//                callbackContext.success();
//            }
//        });
    }

    private void loadBannerAction(JSONArray args, final CallbackContext callbackContext) {
//        final MyTargetAdsPlugin self = this;
//        cordova.getActivity().runOnUiThread(new Runnable() {
//
//            public void run() {
//                if (bannerSizes.get(bannerSize) != null) {
//                    hideBannerView();
//
//                    mBannerAdView = new BannerAdView(self.cordova.getActivity());
//                    mBannerAdView.setAdUnitId(bannerBlockId);
//                    mBannerAdView.setAdSize(bannerSizes.get(bannerSize));
//
//                    final AdRequest adRequest = new AdRequest.Builder().build();
//
//                    mBannerAdView.setBannerAdEventListener(new BannerAdEventListener() {
//                        @Override
//                        public void onAdLoaded() {
//                            bannerLoaded = true;
//                            Log.d(TAG, EVENT_BANNER_DID_LOAD);
//                            self.emitWindowEvent(EVENT_BANNER_DID_LOAD);
//                        }
//
//                        @Override
//                        public void onAdFailedToLoad(@NonNull AdRequestError adRequestError) {
//                            Log.d(TAG, EVENT_BANNER_FAILED_TO_LOAD + ": " + adRequestError.getDescription());
//                            self.emitWindowEvent(EVENT_BANNER_FAILED_TO_LOAD);
//                        }
//
//                        @Override
//                        public void onAdClicked() {
//                            Log.d(TAG, EVENT_BANNER_DID_CLICK);
//                            self.emitWindowEvent(EVENT_BANNER_DID_CLICK);
//                        }
//
//                        @Override
//                        public void onImpression(@Nullable ImpressionData var1) {
//                        }
//
//                        @Override
//                        public void onLeftApplication() {
//
//                        }
//
//                        @Override
//                        public void onReturnedToApplication() {
//
//                        }
//                    });
//
//                    mBannerAdView.loadAd(adRequest);
//                }
//                callbackContext.success();
//            }
//        });
    }

    /**
     * Destroys Yandex Ads Banner and removes it from the container
     */
    private void hideBannerAction(JSONArray args, final CallbackContext callbackContext) {
//        cordova.getActivity().runOnUiThread(new Runnable() {
//            public void run() {
//                hideBannerView();
//                callbackContext.success();
//            }
//        });
    };

    private void hideBannerView() {
//        cordova.getActivity().runOnUiThread(new Runnable() {
//            public void run() {
//                if (mBannerAdView != null) {
//                    if (parentLayout != null && bannerContainerLayout != null) {
//                        if (mBannerAdView.getParent() != null) {
//                            bannerContainerLayout.removeView(mBannerAdView);
//                        }
//                        if (bannerContainerLayout.getParent() != null) {
//                            parentLayout.removeView(bannerContainerLayout);
//                        }
//                    }
//                    destroyBanner();
//                }
//            }
//        });
    }

    /**
     * Destory Banner
     */
    private void destroyBanner() {
//        if (mBannerAdView != null) {
//            mBannerAdView.destroy();
//            mBannerAdView = null;
//        }
    }

    private void setUserConsentAction(Boolean value, final CallbackContext callbackContext) {
        MyTargetPrivacy.setUserConsent(value);
        Log.d(TAG, "setUserConsent: " + value);
        callbackContext.success();
    }
}

File: src/ios/MyTargetAdsPlugin.m
#import "MyTargetAdsPlugin.h"
#import <Cordova/CDV.h>

static NSString *const EVENT_INTERSTITIAL_LOADED = @"interstitialLoaded";
static NSString *const EVENT_INTERSTITIAL_SHOWN = @"interstitialShown";
static NSString *const EVENT_INTERSTITIAL_CLOSED = @"interstitialClosed";
static NSString *const EVENT_INTERSTITIAL_FAILED_TO_LOAD = @"interstitialFailedToLoad";

static NSString *const EVENT_REWARDED_VIDEO_LOADED = @"rewardedVideoLoaded";
static NSString *const EVENT_REWARDED_VIDEO_FAILED = @"rewardedVideoFailed";
static NSString *const EVENT_REWARDED_VIDEO_REWARDED = @"rewardedVideoRewardReceived";
static NSString *const EVENT_REWARDED_VIDEO_STARTED = @"rewardedVideoStarted";
static NSString *const EVENT_REWARDED_VIDEO_CLOSED = @"rewardedVideoClosed";

static NSString *const EVENT_BANNER_DID_LOAD = @"bannerDidLoad";
static NSString *const EVENT_BANNER_FAILED_TO_LOAD = @"bannerFailedToLoad";
static NSString *const EVENT_BANNER_DID_CLICK = @"bannerDidClick";
static NSString *const EVENT_BANNER_WILL_PRESENT_SCREEN = @"bannerWillPresentScreen";
static NSString *const EVENT_BANNER_DID_DISMISS_SCREEN = @"bannerDidDismissScreen";
static NSString *const EVENT_BANNER_WILL_LEAVE_APPLICATION = @"bannerWillLeaveApplication";

@implementation MyTargetAdsPlugin

#pragma mark - CDVPlugin

/**
 * Init
 * @params {CDVInvokedUrlCommand} command
 */
- (void)init:(CDVInvokedUrlCommand *)command
{
    self.rewardedBlockId = [[command argumentAtIndex:0] intValue];
    self.interstitialBlockId = [[command argumentAtIndex:1] intValue];
    self.bannerBlockId = [[command argumentAtIndex:2] intValue];

    self.bannerAtTop = false;
    self.bannerOverlap = false;

//    [MTRGManager setDebugMode:YES];
//    MTRGManager.sdkConfig = [[MTRGConfig.newBuilder withTestDevices:@[@"255992AB-AF6C-499A-910A-23B1A4EE5BA7"]] build];

    NSArray *bannerSizes = @[
        @{
            @"BANNER_320x50": [NSValue valueWithCGSize:(CGSize){320, 50}],
            @"BANNER_320x100": [NSValue valueWithCGSize:(CGSize){320, 100}],
            @"BANNER_300x250": [NSValue valueWithCGSize:(CGSize){300, 250}],
            @"BANNER_300x300": [NSValue valueWithCGSize:(CGSize){300, 300}],
            @"BANNER_240x400": [NSValue valueWithCGSize:(CGSize){240, 400}],
            @"BANNER_400x240": [NSValue valueWithCGSize:(CGSize){400, 240}],
            @"BANNER_728x90": [NSValue valueWithCGSize:(CGSize){728, 90}],
        }
    ];
    NSString* str = nil;
    NSDictionary* options = [command argumentAtIndex:3 withDefault:[NSNull null]];
    str = [options objectForKey:@"bannerAtTop"];
    NSLog(@"%@", str);
    if(str) self.bannerAtTop = [str boolValue];
    str = [options objectForKey:@"overlap"];
    if(str) self.bannerOverlap = [str boolValue];
    str = [options objectForKey:@"bannerSize"];
    if(str) {
        self.bannerSize = [bannerSizes[0][str] CGSizeValue];
    } else {
        self.bannerSize = [bannerSizes[0][@"BANNER_320x50"] CGSizeValue];
    }

    // Send callback successfull
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

/**
 * Emit window event
 * @param {NString} - event name
 */
- (void)emitWindowEvent:(NSString *)event
{
    NSString *js = [NSString stringWithFormat:@"cordova.fireWindowEvent('%@')", event];
    [self.commandDelegate evalJs:js];
}

/**
 * Emits window event with data
 * @param {NSString} - event name
 * @param {NSDictionary} - event data
 */
- (void)emitWindowEvent:(NSString *)event withData:(NSDictionary *)data
{
    NSError *error = nil;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:data options:kNilOptions error:&error];

    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    NSString *js = [NSString stringWithFormat:@"cordova.fireWindowEvent('%@', %@)", event, jsonString];
    [self.commandDelegate evalJs:js];
}

#pragma mark - Rewarded Video Delegate Functions

/**
 * Load rewarded video
 */
- (void)loadRewardedVideo:(CDVInvokedUrlCommand *)command
{
    self.rewardedAd = [MTRGRewardedAd rewardedAdWithSlotId:self.rewardedBlockId];
    self.rewardedAd.delegate = self;

    [self.rewardedAd load];

    // Send callback successfull
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

/**
 * Show rewarded video
 */
- (void)showRewardedVideo:(CDVInvokedUrlCommand *)command
{
    NSLog(@"%s", "showRewardedVideo");
    [self.rewardedAd showWithController:self.viewController];

    // Send callback successfull
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)onLoadWithRewardedAd:(MTRGRewardedAd *)rewardedAd
{
    NSLog(@"%s", __PRETTY_FUNCTION__);
    [self emitWindowEvent:EVENT_REWARDED_VIDEO_LOADED];
}

- (void)onNoAdWithReason:(NSString *)reason rewardedAd:(MTRGRewardedAd *)rewardedAd
{
    NSLog(@"rewardedAdDidFailToLoad");
    NSLog(@"%@", reason);
    [self emitWindowEvent:EVENT_REWARDED_VIDEO_FAILED];
}

- (void)onDisplayWithRewardedAd:(MTRGRewardedAd *)rewardedAd
{
    [self emitWindowEvent:EVENT_REWARDED_VIDEO_STARTED];
}

- (void)onReward:(MTRGReward *)reward rewardedAd:(MTRGRewardedAd *)rewardedAd
{
    [self emitWindowEvent:EVENT_REWARDED_VIDEO_REWARDED];
}

- (void)onCloseWithRewardedAd:(MTRGRewardedAd *)rewardedAd
{
    [self emitWindowEvent:EVENT_REWARDED_VIDEO_CLOSED];
}


- (void)loadBanner:(CDVInvokedUrlCommand *)command
{
    NSLog(@"loadBanner");
    NSLog(@"%@", self.bannerBlockId);
//    if (self.adView != nil) {
//        [self destroyBanner];
//    }
//    YMAAdSize *adSize = [YMAAdSize fixedSizeWithCGSize:self.bannerSize];
//    self.adView = [[YMAAdView alloc] initWithAdUnitID:self.bannerBlockId adSize:adSize];
//    self.adView.delegate = self;
//    [self.adView loadAd];
}

#pragma mark - Banner Delegate Functions

// Show banner
- (void)showBanner:(CDVInvokedUrlCommand *)command
{
//    if(self.adView)
//    {
//        UIView* parentView = self.bannerOverlap ? self.viewController.view : [self.viewController.view superview];
//
//        if (self.bannerAtTop) {
//            [self.adView displayAtTopInView:parentView];
//        } else {
//            [self.adView displayAtBottomInView:parentView];
//        }
//    }
//
//    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
//    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)hideBanner:(CDVInvokedUrlCommand *)command
{
//    NSLog(@"%s", "hideBanner");
//    [self destroyBanner];
}

- (void)destroyBanner
{
//    [self.adView setDelegate:nil];
//    [self.adView removeFromSuperview];
//    self.adView = nil;
}

//- (void)adViewDidFailLoading:(YMAAdView *)bannerAdView error:(NSError *)error {
//    NSLog(@"%s", "adViewDidFailLoading");
//    NSLog(@"%@", error.description);
//    [self emitWindowEvent:EVENT_BANNER_FAILED_TO_LOAD];
//}
//
//- (void)adViewDidLoad:(YMAAdView *)adView {
//    NSLog(@"%s", "adViewDidLoad");
//    [self emitWindowEvent:EVENT_BANNER_DID_LOAD];
//}
//
//- (void)adViewDidClick:(YMAAdView *)adView {
//    NSLog(@"%s", "adViewDidClick");
//    [self emitWindowEvent:EVENT_BANNER_DID_CLICK];
//}
//
//
//- (void)adViewWillLeaveApplication:(YMAAdView *)adView {
//    NSLog(@"%s", __PRETTY_FUNCTION__);
//    [self emitWindowEvent:EVENT_BANNER_WILL_LEAVE_APPLICATION];
//}
//
//- (void)adView:(YMAAdView *)adView willPresentScreen:(nullable UIViewController *)viewController {
//    NSLog(@"%s", @"willPresentScreen");
//    [self emitWindowEvent:EVENT_BANNER_WILL_PRESENT_SCREEN];
//}

#pragma mark - Intersitial Delegate Functions

- (void)loadInterstitial:(CDVInvokedUrlCommand *)command
{
    self.interstitialAd = [MTRGInterstitialAd interstitialAdWithSlotId:self.interstitialBlockId];
    self.interstitialAd.delegate = self;

    [self.interstitialAd load];

    // Send callback successfull
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)showInterstitial:(CDVInvokedUrlCommand *)command
{
    [self.interstitialAd showWithController:self.viewController];

    // Send callback successfull
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)onLoadWithInterstitialAd:(MTRGInterstitialAd *)interstitialAd
{
    [self emitWindowEvent:EVENT_INTERSTITIAL_LOADED];
}

- (void)onNoAdWithReason:(NSString *)reason interstitialAd:(MTRGInterstitialAd *)interstitialAd
{
    NSLog(@"interstitialAdDidFailToLoad");
    NSLog(@"%@", reason);
    [self emitWindowEvent:EVENT_INTERSTITIAL_FAILED_TO_LOAD];
}

- (void)onDisplayWithInterstitialAd:(MTRGInterstitialAd *)interstitialAd
{
    [self emitWindowEvent:EVENT_INTERSTITIAL_SHOWN];
}

- (void)onCloseWithInterstitialAd:(MTRGInterstitialAd *)interstitialAd
{
    [self emitWindowEvent:EVENT_INTERSTITIAL_CLOSED];
}

- (void)setUserConsent:(CDVInvokedUrlCommand *)command {
    [MTRGPrivacy setUserConsent:[[command argumentAtIndex:0] boolValue]];
    NSLog(@"setUserConsent: ");
    NSLog([[command argumentAtIndex:0] boolValue] ? @"YES" : @"NO");
}

@end
